### PR TITLE
Extend SGR and VT200 (normal) mouse support (fixes #714, #506)

### LINKED ISSUE #506: Right Mouse Click does not work in ncmpcpp
In [ncmpcpp](https://rybczak.net/ncmpcpp/) you can use the right mouse button to play a track directly. This works in `urxvt` and `gnome-terminal`, but not in `alacritty` - nothing happens if I right-click on a track.

Left clicking (select a track) and scrolling using the mouse wheel works though.

```
$ cat /etc/issue
Arch Linux \r (\l)

$ alacritty -V
alacritty 0.1.0

$ ncmpcpp -v
ncmpcpp 0.7.7
optional screens compiled-in:
 - tag editor
 - tiny tag editor
 - artist info
 - outputs
 - visualizer
 - clock
encoding detection: enabled
built with support for: curl fftw ncurses taglib unicode

$ pacman -Qi ncurses | head -n 4
Name            : ncurses
Version         : 6.0+20170128-1
Description     : System V Release 4.0 curses emulation library
Architecture    : x86_64
```

### LINKED ISSUE #714: Emacs scrolls up by itself on mouse click or drag
Emacs scrolls up by itself when navigating cursor with mouse click or drag.
You might reproduce immediately on bare emacs (25 or 26)  by 
```sh
emacs -nw --no-init-file --no-loadup --eval '(xterm-mouse-mode t)' some-long-file
```
Vim doesn't with ```set mouse=a```, and Emacs with same flag on gnome-terminal, guake, xterm doesn't.

I built Alacritty using rustc 1.21.0-nightly (1d2a6df38 2017-08-03), Ubuntu 17.04.

### LINKED ISSUE #1145: Mouse modifiers should be propagated
Currently alacritty does not propagate modifiers for mouse reports. Some applications might make use of the ctrl or alt modifier, this should be propagated to the application following the xterm implementation.

Propagating shift should not be necessary because shift stops alacritty from propagating the mouse to the application.

An application that makes use of this is `kakoune`.

### PR DESCRIPTION
**Pull request by Celti**
This PR extends Alacritty's mouse handling to handle:
* Reporting all three of the left, right, and middle mouse buttons.
* Reporting mouse dragging events with all three buttons.
* Reporting mouse movement events when no button is pressed.

This PR does *not* report modifier keys with clicks/drags, because I'm unsure what modifiers (if any) we want or need to handle. We *can't* report Shift, as that is the de-facto standard modifier for ignoring application mouse handling (and Alacritty already treats it as such). The vast majority of CLI applications out there don't *expect* Ctrl to be reported, because xterm mandates Ctrl-Click to access its menus. They also do not generally expect Alt, as the de-facto standard modifier for moving and resizing windows with the mouse. The Windows key, of course, is also generally ignored.

This PR also does not support more than three mouse buttons (plus the wheel). While Glutin reports them, VT200/SGR button codes treat 0 as left, 1 as middle, 2 as right, and 3 as either "button release" (VT200), "any other button" (SGR), or "no button pressed" (SGR with any-event tracking); thus, we skip supporting them entirely.

This fixes #714, fixes #506 and fixes #1145.

### CODE REVIEW

File: src/input.rs [hunk 1]
```diff
@@ -466,18 +492,49 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
...
+                let state = mem::replace(&mut self.ctx.mouse_mut().right_button_state, state);
+                if self.ctx.mouse_mut().right_button_state != state {
+                    match self.ctx.mouse_mut().right_button_state {
+                        ElementState::Pressed => {
+                            self.on_mouse_press(button, modifiers);
+                        },
+                        ElementState::Released => {
+                            self.on_mouse_release(button, modifiers);
+                        }
+                    }
```
Comment (chrisduerr) [inline src/input.rs]: It looks to me like the matches inside the state change check all do the same. Might be cleaner to refactor this to update the state of the button when it changed and then handle the rest for all of them at once.

Just an idea:
```rust
let button_state = match button {
  MouseButton::Left => Some(&mut self.ctx.mouse_mut().left_button_state),
  MouseButton::Middle => Some(&mut self.ctx.mouse_mut().middle_button_state),
  ...
  _ => None
};

if let Some(button_state) = button_state && button_state != state {
  button_state = state;
  match button_state {
    ElementState::Pressed => self.on_mouse_press(button, modifiers);
    ...
  }
}
```

File: src/input.rs [hunk 2]
```diff
@@ -466,18 +492,49 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
...
+                }
+            }
+            MouseButton::Other(_) => {
+                // Handling more than three buttons properly is impossible.
+            }
```
Comment (chrisduerr) [inline src/input.rs]: Why is that impossible? I don't think we need more than three buttons, but it seems possible to me.

Also changing this to
```rust
// Handling more than three buttons properly is impossible.
MouseButton::Other(_) => (),
```
would be the better choice, I think.

File: src/input.rs [hunk 3]
```diff
@@ -362,9 +378,14 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
...
                 self.ctx.clear_selection();
-                let report_modes = mode::TermMode::MOUSE_REPORT_CLICK | mode::TermMode::MOUSE_MOTION;
+                let report_modes = mode::TermMode::MOUSE_REPORT_CLICK | mode::TermMode::MOUSE_DRAG | mode::TermMode::MOUSE_MOTION;
                 if !modifiers.shift && self.ctx.terminal_mode().intersects(report_modes) {
-                    self.mouse_report(0, ElementState::Pressed);
+                    match button {
+                        MouseButton::Left   => self.mouse_report(0, ElementState::Pressed),
+                        MouseButton::Middle => self.mouse_report(1, ElementState::Pressed),
+                        MouseButton::Right  => self.mouse_report(2, ElementState::Pressed),
+                        MouseButton::Other(_) => { } // Can't handle more than three buttons.
```
[the remark this answers is not in the document]
Comment (Celti) [reply src/input.rs]: Noted, I'll fix this and the others.

File: src/input.rs (hunk 1 above)
Comment (Celti) [reply src/input.rs]: I knew there was a lot of duplication there, that way of refactoring just didn't occur to me! I'll get right on that.

File: src/input.rs (hunk 2 above)
Comment (Celti) [reply src/input.rs]: Can fix the comment; as for handling more than three buttons:
* VT200 (i.e., what alacritty calls "normal" mode) has no concept of more than three buttons plus two wheel directions.
* SGR mode lumps all buttons past the third into a single button code of "3".
* VT200 mode uses button code "3" for button-release events.
* Both VT200 and SGR mode use button code 3 to represent "no button pressed" for any-event motion tracking mode.

With all of those, you can properly support *one* extra button, only in SGR mode, and only if you don't use any-event motion tracking. That's close enough to impossible for me.

**Comment by chrisduerr** [comment]
Looks like github is incorrectly detecting some file changes, but it should be fine.

File: src/input.rs [hunk 4]
```diff
@@ -466,17 +494,20 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
...
-            if self.ctx.mouse_mut().left_button_state != state {
-                match self.ctx.mouse_mut().left_button_state {
-                    ElementState::Pressed => {
-                        self.on_mouse_press(modifiers);
-                    },
-                    ElementState::Released => {
-                        self.on_mouse_release(modifiers);
-                    }
-                }
+        let button_state = match button {
```
Comment (chrisduerr) [inline src/input.rs]: Might be good to change this to `old_state` (or something like that) to clearly indicate that this is the result of the `mem::replace` and it is neither the current `_button_state` nor the current `state`.

Comment (Celti) [reply src/input.rs]: Agreed. Updated to `prev_state` to match other uses of `mem::replace` in the file.

**Comment by chrisduerr** [comment]
@Celti It seems I've accidentally implemented this PR again in #1141, but there seem to be some advantages of both. Mine is definitely missing the selection dragging and I don't think you've implemented the modifiers.

The main question I have when comparing the two versions is that I don't understand why you have a middle and right button state. From what I understand the left button state is only set so selection can be updated, I don't see any part of the code that reads from middle/right button state.

**Comment by chrisduerr** [comment]
@Celti I've attempted to merge our two PRs and directly pushed to your branch in an attempt to reduce the waiting times a bit.

If you have the time please feel free to review this and leave some comments about the changes I've made.

**Comment by Celti** [comment]
Sounds good! Exhausted right now, I'll look it over when I wake up tonight.

**Comment by jwilm** [comment]
I thought we had merged this! My bad. Will help to prioritize this after @Celti gets a chance to review the additional changes.

File: src/input.rs [hunk 5]
```diff
@@ -338,13 +337,26 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
...
-    pub fn mouse_report(&mut self, button: u8, state: ElementState) {
-        if self.ctx.terminal_mode().contains(mode::TermMode::SGR_MOUSE) {
-            self.sgr_mouse_report(button, state);
+    pub fn mouse_report(&mut self, button: u8, state: ElementState, modifiers: ModifiersState) {
+        // Calculate modifiers value
+        let mut mods = 0;
+        if modifiers.shift {
+            mods += 4;
+        }
+        if modifiers.logo {
```
Comment (Celti) [inline src/input.rs]: The protocol specifies 8 is Meta (Alt, in Glutin terms) — Glutin's Logo modifier is X11's Super or Hyper.

**Comment by Celti** [comment]
Sorry for the long delay — I needed more sleep than I thought!

Everything looks good, and tests out just fine, with the small issue that this patch reports the logo key as Meta/Alt. Is that deliberate?

File: src/input.rs (hunk 5 above)
Comment (chrisduerr) [reply src/input.rs]: You're right, alt probably makes more sense here. Meta is just super for me :P.

**Comment by chrisduerr** [comment]
@Celti Do you know what we would have to do to make Ctrl+Click work? I think we would have to adjust the terminfo. I can't get `infocmp` to work properly so I wasn't able to modify `xterm-256color`. But I've noticed that we could just copy `rxvt-unicode-256color` and Ctrl+Click would work.

**Comment by Celti** [comment]
Terminfo does not specify anything regarding mice and modifier keys; as long as `BT`, `Gm`, `Km`, `Mi`, and `RQ` are all set properly (and the `xterm-256color` entry that `alacritty`'s entry builds off of has them set appropriately), support for Ctrl-Click *should* be entirely dependent on the application running inside Alacritty.

**Comment by chrisduerr** [comment]
`kakoune` seems to support Ctrl+Click. urxvt and kitty (the two I've tested that support this), seem to prefer normal mode over SGR mode, maybe that's why. Using the terminfo from urxvt, `kakoune` uses normal mouse reports and Ctrl+Click works.

Maybe there's an issue with SGR mode and that's why urxvt uses normal mode?

**Comment by chrisduerr** [comment]
`xterm` has their popup menu on ctrl-click by default. That's why I assumed the terminfo might not support it.

**Comment by Celti** [comment]
> Maybe there's an issue with SGR mode and that's why urxvt uses normal mode?
urxvt actually doesn't use normal mode *or* SGR mode (at least as of a few years ago, there were patches to make it support SGR mode, I don't know if they've been mainlined) — It uses a weird Unicode superset of normal mode.

> `xterm` has their popup menu on ctrl-click by default. That's why I assumed the terminfo might not support it.
Which is a good thought, but as it turns out terminfo doesn't care one way or another about it, it just wants to know what escape sequences that indicate a mouse command *look* like, not whether you'll actually pass all of them or only a subset through.

**Comment by chrisduerr** [comment]
I didn't mean that urxvt uses normal mode, but that alacritty uses normal mode with the urxvt terminfo. So my thought was that there's some reason why SGR mode can not provide the control clicks.

Manually requesting SGR mode with urxvt/kitty and comparing the output looks like this PR handles SGR mode correctly, so I don't think the implementation is wrong.

**Pull request merged**